add Async & Approval config to webhook action

andrewstillv15 opened the pull request:
Resolves #

### Problem

<!---
  Describe the problem this PR is solving. What is the application state
  before this PR is merged?
-->

### Solution

<!---
  Describe the way this PR solves the above problem. Add as much detail as you
  can to help reviewers understand your changes. Include any alternatives and
  tradeoffs you considered.

  For Terraform you'll want to make sure we can CRUD the resource and then also
  be able to mutate different fields with different values, especially null.

### Given this Terraform config file
```tf

```

### The `terraform apply` output
```bash

```
-->

### Checklist

- [ ] I have run this code, and it appears to resolve the stated issue.
- [ ] This PR does not reduce total test coverage
- [ ] This PR has no user interface changes or has already received approval from product management to change the interface.
- [ ] Make a [changie](https://github.com/OpsLevel/terraform-provider-opslevel/blob/main/CONTRIBUTING.md#changie-change-log-generation) entry that explains the customer facing outcome of this change

rocktavious commented on opslevel/resource_opslevel_webhook_action.go:
```diff
@@ -63,6 +65,10 @@ func (r *WebhookActionResource) Schema(ctx context.Context, req resource.SchemaR
 		Attributes: map[string]schema.Attribute{
+			"async": schema.BoolAttribute{
+				MarkdownDescription: "Whether or not the Webhook Action expects an additional, asynchronous response upon completion.",
+				Required:            true,
```
So this is the most contentious part.  As is this means when customers upgrade it'll be a breaking change so we'll have to do a minor version bump.

If we want it to not be a breaking change we could modify it to be

```suggestion
				Optional:            true,
				Default:             booldefault.StaticBool(false),
```

This should preserve how it worked prior to this change and not cause customers existing configuration to blow up once they upgrade provider versions.  The trade off is - does this work properly down in Create and Update.

andrewstillv15 replied:
This is actually more true to the create & update behaviour too

rocktavious commented on opslevel/resource_opslevel_webhook_action.go:
```diff
@@ -63,6 +65,10 @@ func (r *WebhookActionResource) Schema(ctx context.Context, req resource.SchemaR
 		Attributes: map[string]schema.Attribute{
+			"async": schema.BoolAttribute{
+				MarkdownDescription: "Whether or not the Webhook Action expects an additional, asynchronous response upon completion.",
+				Required:            true,
```
So this is the most contentious part.  As is this means when customers upgrade it'll be a breaking change so we'll have to do a minor version bump.

If we want it to not be a breaking change we could modify it to be

```suggestion
				Optional:            true,
				Default:             booldefault.StaticBool(false),
```

This should preserve how it worked prior to this change and not cause customers existing configuration to blow up once they upgrade provider versions.  The trade off is - does this work properly down in Create and Update.

rocktavious reviewed (commented):
Overall though i think you got this right on the money.

Please make sure you add a changie - describing which field you added to which resource so customers know.  The changelogs in the tools are much more highly watched then for the monolith.

rocktavious commented on opslevel/resource_opslevel_trigger_definition.go:
```diff
@@ -106,6 +114,48 @@ func (r *TriggerDefinitionResource) Schema(ctx context.Context, req resource.Sch
+							Computed:    true,
+							Default:     booldefault.StaticBool(false),
+						},
+						"teams": schema.ListNestedAttribute{
+							Description: "Teams that can approve.",
+							Optional:    true,
+							NestedObject: schema.NestedAttributeObject{
+								Attributes: map[string]schema.Attribute{
+									"alias": schema.StringAttribute{
+										Required: true,
+									},
+									"id": schema.StringAttribute{
+										Required: true,
+									},
+								},
+							},
+						},
```
The way we've handled "IdentifierInput" is that we have a special function in opslevel-go that can make a new `IdentifierInput` based on if the string is detected as an ID - then it fills out the ID field - and if not it fills out the alias field.  This means the customer can just use a list of strings like this
```tf
  teams = ["098sdflkjwerlkjwertkljsdf90sdf908", "my-team-alias", "my-team-alias2"]
```

So i think this can just be a `ListAttribute` of strings.  This provides a much better UX to the customer.

Same goes for the `users` field.


Where this gets tricky on our side is that we need to shave into the state exactly what they gave us instead of what our API returned.  We usually do this as special case logic in the  "NewModel" function.

rocktavious commented on opslevel/resource_opslevel_trigger_definition.go:
```diff
@@ -106,6 +114,48 @@ func (r *TriggerDefinitionResource) Schema(ctx context.Context, req resource.Sch
 				Required:    true,
 				Validators:  []validator.String{IdStringValidator()},
 			},
+			"approval_config": schema.MapNestedAttribute{
```
Generally you want to avoid complex object types like this in terraform and just lift the fields up to the parent.  If you feel strongly about keeping the Nested map.  Then we should adjust the module to instead of having to do

```
variable "approval_config" {
  type = map(object({
    approval_required = bool
    teams = list(object({
      alias = string
      id    = string
    }))
    users = list(object({
      email = string
      id    = string
    }))
  }))
  default = null
}
```

we can just do

```
variable "approval_required" {
  type = bool
  default = false
}
variable "approval_teams" {
  type = list(string)
default = []
}
variable "approval_users" {
  type = list(string)
  default = []
}
```

and map those into the object like
```tf
resource "opslevel_trigger_definition" "test" {
  access_control           = var.access_control
  action                   = var.action
  approval_config          = {
    approval_required = var.approval_required
    teams = var.approval_teams
    users = var.approval_users
  }
  description              = var.description
  entity_type              = var.entity_type
```

rocktavious commented on opslevel/resource_opslevel_trigger_definition.go:
```diff
@@ -36,18 +37,25 @@ type TriggerDefinitionResource struct {
-	Published              types.Bool   `tfsdk:"published"`
+	AccessControl          types.String    `tfsdk:"access_control"`
+	Action                 types.String    `tfsdk:"action"`
+	ApprovalConfig         *ApprovalConfig `tfsdk:"approval_config"`
```
Is the pointer required here?

rocktavious commented:
@andrewstillv15 - sorry just realizing this now after doing a more through review.

We have example terraform resources that get pulled into our docs.  We should probably update them for these new fields.

https://github.com/OpsLevel/terraform-provider-opslevel/blob/main/examples/resources/opslevel_trigger_definition/resource.tf
and
https://github.com/OpsLevel/terraform-provider-opslevel/blob/main/examples/resources/opslevel_webhook_action/resource.tf

For the async field you can probably do something like this
```tf
  method      = "POST"
  #async = true #Uncomment if your action is long running
  headers = {
```
For the approval config it might be handy to duplicate the example and show a 2 different ways to configure it.

Terraform users will often look to the examples first by copy &  paste and then only if they have a problem will they actually read the documentation 😆

rocktavious commented:
@andrewstillv15 ok i've found 1 bug so far and here is the flow.

Start with
```
  approval_required        = true
  approval_users           = ["kyle.rockman@mac.com"]
```
then apply
then in the UI modify the approval teams to have a team
then modify terraform configuration to
```
  approval_required        = true
  approval_teams           = []
  approval_users           = ["kyle.rockman@mac.com"]
```
then apply
notice in the UI the approval teams is still present
run apply again - notice terraform reports "no changes"

I suspect this will be the case for users too.  Its a tricky logic bug in UPDATE most likely where it needs to work slightly different then for CREATE.

rocktavious commented:
Found another bug - this one is likely because of the API - but we might need to code around it.

After going the above flow i changed terraform to
```
  approval_required        = false
  approval_teams           = []
  approval_users           = ["kyle.rockman@mac.com"]
```
and then got the following terraform bug report

```
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to opslevel_trigger_definition.example, provider "provider[\"registry.terraform.io/opslevel/opslevel\"]" produced an unexpected new value: .approval_users: was
│ cty.ListVal([]cty.Value{cty.StringVal("kyle.rockman@mac.com")}), but now null.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.

```

andrewstillv15 commented:
@rocktavious I've managed a fix for both bugs. The fix for this 

> Found another bug - this one is likely because of the API - but we might need to code around it.
> 
> After going the above flow i changed terraform to
> 
> ```
>   approval_required        = false
>   approval_teams           = []
>   approval_users           = ["kyle.rockman@mac.com"]
> ```
> 
> and then got the following terraform bug report
> 
> ```
> │ Error: Provider produced inconsistent result after apply
> │ 
> │ When applying changes to opslevel_trigger_definition.example, provider "provider[\"registry.terraform.io/opslevel/opslevel\"]" produced an unexpected new value: .approval_users: was
> │ cty.ListVal([]cty.Value{cty.StringVal("kyle.rockman@mac.com")}), but now null.
> │ 
> │ This is a bug in the provider, which should be reported in the provider's own issue tracker.
> ```


was to set a default value for both the teams and the users to be an empty array to empower the state to emulate the API behaviour (neither of the approval teams or users will ever be set to nil, if set to nil they will return an empty array).


The fix for this

> @andrewstillv15 ok i've found 1 bug so far and here is the flow.
> 
> Start with
> 
> ```
>   approval_required        = true
>   approval_users           = ["kyle.rockman@mac.com"]
> ```
> 
> then apply then in the UI modify the approval teams to have a team then modify terraform configuration to
> 
> ```
>   approval_required        = true
>   approval_teams           = []
>   approval_users           = ["kyle.rockman@mac.com"]
> ```
> 
> then apply notice in the UI the approval teams is still present (which is shouldn't be because we've set the terraform be explictly an `[]`) run apply again - notice terraform reports "no changes" (this is also bad it means we are writing into the state the wrong data)
> 
> I suspect this will be the case for users too. Its a tricky logic bug in UPDATE most likely where it needs to work slightly different then for CREATE.

was to update how the local state was set. The previous behaviour would only copy items to the new state if there was a match with the existing state (which obviously would stop remote resource updates from being reflected). I've also modified the decision to return an empty list to be solely dependent on the remote state as it will always be the source of truth.

rocktavious commented:
Ok so here's another one i found

```
│ Unable to create trigger definition, got error: OpsLevel API Errors:
│       - 'approvalRequired' When approval is required, there must be at least one User or Team assigned.

```

Generally we try to add validation on the terraform side so that things like this in the API show up during `terraform plan` before the customer merges to main and runs `terraform apply`.

I won't die on this hill if its not straight forward to implement.

rocktavious commented:
Alright - i've done a crap ton of manual testing and other then the nitty validation issue (which is just  quality of life thing) this is great!!!

andrewstillv15 commented:
> Ok so here's another one i found

│ Unable to create trigger definition, got error: OpsLevel API Errors:
│       - 'approvalRequired' When approval is required, there must be at least one User or Team assigned.

Generally we try to add validation on the terraform side so that things like this in the API show up during terraform plan before the customer merges to main and runs terraform apply.

I won't die on this hill if its not straight forward to implement.

Spent about 2 hours on this, couldn't figure it out.
